com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/searcher/TestItemSearcher.java

```diff
@@ -0,0 +1,113 @@
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package com.microsoft.java.test.plugin.searcher;
+
+import com.microsoft.java.test.plugin.model.TestItem;
+import com.microsoft.java.test.plugin.model.TestKind;
+import com.microsoft.java.test.plugin.model.TestLevel;
+import com.microsoft.java.test.plugin.util.TestSearchUtils;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IMethod;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.search.IJavaSearchConstants;
+import org.eclipse.jdt.core.search.IJavaSearchScope;
+import org.eclipse.jdt.core.search.SearchEngine;
+import org.eclipse.jdt.core.search.SearchParticipant;
+import org.eclipse.jdt.core.search.SearchPattern;
+import org.eclipse.jdt.core.search.SearchRequestor;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.lsp4j.Range;
+
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+
+@SuppressWarnings("restriction")
+public abstract class TestItemSearcher {
+    protected final SearchEngine searchEngine;
+    protected SearchParticipant[] searchParticipants;
+    protected int searchFor;
+    protected TestLevel searchResultLevel;
+
+    public TestItemSearcher(int searchFor, TestLevel searchResultLevel) {
+        this.searchEngine = new SearchEngine();
+        this.searchParticipants = new SearchParticipant[] { SearchEngine.getDefaultSearchParticipant() };
+        this.searchFor = searchFor;
+        this.searchResultLevel = searchResultLevel;
+    }
+
+    public List<TestItem> search(String uri, String fullyqualifiedName, IProgressMonitor monitor)
+            throws JavaModelException, CoreException, URISyntaxException {
+        final List<TestItem> entryList = new ArrayList<>();
+        this.searchEngine.search(this.resolveSearchPattern(fullyqualifiedName), this.searchParticipants,
+                this.resolveSearchScope(uri), this.resolveSeartchRequestor(entryList, fullyqualifiedName), monitor);
+        return entryList;
+    }
+
+    protected SearchPattern resolveSearchPattern(String fullyqualifiedName) {
+        return SearchPattern.createPattern("*", this.searchFor, IJavaSearchConstants.DECLARATIONS,
+                SearchPattern.R_PATTERN_MATCH);
+    }
+
+    protected abstract SearchRequestor resolveSeartchRequestor(List<TestItem> entryList, String fullyqualifiedName);
+
+    protected abstract IJavaSearchScope resolveSearchScope(String uri) throws JavaModelException, URISyntaxException;
+
+    protected TestItem parseSearchItems(IJavaElement element) throws JavaModelException {
```

Comment:
`parseSearchItems` -> `constructTestItem`

Reply:
And looks like these functions (construct test item) could be utility methods inside `TestSearchUtilis`.